starlight/utils/search.py

```diff
@@ -0,0 +1,365 @@
+
+    # sepcial case single attribute
+    if len(attrs) == 1:
+        k, v = attrs.popitem()
+        predicate = v.filter if isinstance(v, SearchFilter) else _eq_predicate(v)
+        unsorted_items_gen = ((item, _get_score(item, k, predicate)) async for item in iterable)
+        unsorted_items = [t async for t in unsorted_items_gen if t[1]]
+        if _sort:
+            unsorted_items.sort(key=lambda t: t[1], reverse=True)
+        items = [t[0] for t in unsorted_items]
+    else:
+        predicates = {k: v.filter if isinstance(v, SearchFilter) else _eq_predicate(v) for k, v in attrs.items()}
+        unsorted_items = []
+
+        if _sort:
+            if _check_any:
+                # any, sorted
+                items_gen = (
+                    (item, [_get_score(item, k, pred) for k, pred in predicates.items()])
+                    async for item in iterable
+                )
+                unsorted_items = [t async for t in items_gen if _check(t[1])]
+            else:
+                # need the scores to sort, but try to short circuit if it fails
+                num_predicates = len(predicates)
+                # all, sorted
+                async for item in iterable:
+                    scores = []
+                    for k, pred in predicates.items():
+                        score = _get_score(item, k, pred)
+                        if not score:
+                            break
+                        scores.append(score)
+                    if len(scores) == num_predicates:
+                        unsorted_items.append((item, scores))
+
+            unsorted_items.sort(key=lambda t: t[1])
+            items = [t[0] for t in unsorted_items]
+        else:
+            # not sorted, don't need the score
+            items = [
+                item
+                async for item in iterable
+                if _check(_get_score(item, k, pred) for k, pred in predicates.items())
+            ]
+    return items
+
+
+@overload
+def search(iterable: AsyncIterable[T], /, *, check_any: bool = False, sort: bool = False,  **attrs: Any) -> Coro[List[T]]:
+    ...
+
+@overload
+def search(iterable: Iterable[T], /, *, check_any: bool = False, sort: bool = False,  **attrs: Any) -> List[T]:
+    ...
+
+def search(
+    iterable: _Iter[T], /, *, check_any: bool = False, sort: bool = False, **attrs: Any
+) -> Union[List[T], Coro[List[T]]]:
+    r"""A helper that returns all elements in the iterable that meet all or any
+    of the traits passed in ``attrs`` depending on ``check_any``.
+    To have a nested attribute search (i.e. search by ``x.y``) then
+    pass in ``x__y`` as the keyword argument.
+    If nothing is found that matches the attributes passed, then
+    an empty list is returned.
+
+    Examples
+    ---------
+    Using :class:`SearchFilter`:
+    .. code-block:: python3
+        cmds_by_name_and_desc = await starlight.search(bot.commands, name=FuzzyFilter('user'), description=ContainsFilter('user'))
```

Comment:
Add a new blank line between `.. code-block:: python3` and the code 
```suggestion

cmds_by_name_and_desc = await starlight.search(bot.commands, name=FuzzyFilter('user'), description=ContainsFilter('user'))
```
Applies for all codeblocks

Reply:
did it fix